src/renderer/TooltipRenderer.tsx

```diff
@@ -0,0 +1,383 @@
+import { Rectangle } from '../shape/Rectangle';
+import { getRadialCursorPoints } from '../util/cursor/getRadialCursorPoints';
+import { Sector } from '../shape/Sector';
+import { filterProps } from '../util/ReactUtils';
+import { getCursorRectangle } from '../util/cursor/getCursorRectangle';
+import { getCursorPoints } from '../util/cursor/getCursorPoints';
+import {
+  DefaultTooltipContent,
+  NameType,
+  Payload,
+  ValueType,
+  Props as TooltipContentProps,
+} from '../component/DefaultTooltipContent';
+import { isNumber } from '../util/DataUtils';
+import { ContentType } from '../component/Tooltip';
+
+type Props = {
+  activeCoordinate: ChartCoordinate;
+  chartName: string;
+  layout: LayoutType;
+  offset: ChartOffset;
+  viewBox: CartesianViewBox;
+  tooltipAxisBandSize: number;
+  activeTooltipIndex: number;
+  activeLabel: string;
+  activePayload: Array<Payload<any, any>>;
+};
+
+function CursorRenderer({
+  activeCoordinate,
+  chartName,
+  layout,
+  offset,
+  tooltipAxisBandSize,
+  activeTooltipIndex,
+  activeLabel,
+  activePayload,
+}: Props) {
+  const [context] = useContext(ChartContext);
+  const { isTooltipActive, tooltipEventType, cursor } = context;
+
+  if (!isTooltipActive || !activeCoordinate || (chartName !== 'ScatterChart' && tooltipEventType !== 'axis')) {
+    return null;
+  }
+
+  let restProps;
+  let cursorComp: React.ComponentType<any> = Curve;
+
+  if (chartName === 'ScatterChart') {
+    restProps = activeCoordinate;
+    cursorComp = Cross;
+  } else if (chartName === 'BarChart') {
+    restProps = getCursorRectangle(layout, activeCoordinate, offset, tooltipAxisBandSize);
+    cursorComp = Rectangle;
+  } else if (layout === 'radial') {
+    const { cx, cy, radius, startAngle, endAngle } = getRadialCursorPoints(activeCoordinate);
+    restProps = {
+      cx,
+      cy,
+      startAngle,
+      endAngle,
+      innerRadius: radius,
+      outerRadius: radius,
+    };
+    cursorComp = Sector;
+  } else {
+    restProps = { points: getCursorPoints(layout, activeCoordinate, offset) };
+    cursorComp = Curve;
+  }
+  const cursorProps = {
+    viewBox: { ...offset, x: offset.left, y: offset.top },
+    stroke: '#ccc',
+    pointerEvents: 'none',
+    ...offset,
+    ...restProps,
+    ...filterProps(cursor),
+    payloadIndex: activeTooltipIndex,
+    className: 'recharts-tooltip-cursor',
+    active: isTooltipActive,
+    label: activeLabel,
+    payload: isTooltipActive ? activePayload : [],
+    coordinate: activeCoordinate,
+  };
+
+  return isValidElement(cursor) ? cloneElement(cursor, cursorProps) : createElement(cursorComp, cursorProps);
+}
+
+type UniqueFunc<TValue extends ValueType, TName extends NameType> = (entry: Payload<TValue, TName>) => unknown;
+type UniqueOption<TValue extends ValueType, TName extends NameType> = boolean | UniqueFunc<TValue, TName>;
+
+function defaultUniqBy<TValue extends ValueType, TName extends NameType>(entry: Payload<TValue, TName>) {
+  return entry.dataKey;
+}
```

Comment:
this is defined in the file above too?

Reply:
I don't actually intend to merge this the way it is now. All defaults resolving should be happening before moving data to context. So the private renderer can only work with one shape of props.